rust/src/tmpfiles.rs

```diff
@@ -0,0 +1,122 @@
+
+use anyhow::{Context, Result};
+use cap_std::fs::{Dir, Permissions};
+use cap_std_ext::dirext::CapStdExtDirExt;
+use fn_error_context::context;
+use std::collections::{HashMap, HashSet};
+use std::fmt::Write;
+use std::os::unix::prelude::PermissionsExt;
+use std::path::Path;
+
+const TMPFILESD: &str = "usr/lib/tmpfiles.d";
+const RPMOSTREE_TMPFILESD: &str = "usr/lib/rpm-ostree/tmpfiles.d";
+const AUTOVAR_PATH: &str = "rpm-ostree-autovar.conf";
+
+#[context("Deduplicate tmpfiles entries")]
+pub fn deduplicate_tmpfiles_entries(tmprootfs_dfd: i32) -> CxxResult<()> {
+    let tmprootfs_dfd = unsafe { ffiutil::ffi_dirfd(tmprootfs_dfd)? };
+
+    // scan all rpm-ostree auto generated entries and save
+    let tmpfiles_dir = tmprootfs_dfd
+        .open_dir(RPMOSTREE_TMPFILESD)
+        .context("readdir {RPMOSTREE_TMPFILESD}")?;
+    let mut rpmostree_tmpfiles_entries = save_tmpfile_entries(&tmpfiles_dir)?
+        .map(|s| {
+            let entry = tmpfiles_entry_get_path(&s.as_str())?;
+            anyhow::Ok((entry.to_string(), s.to_string()))
+        })
+        .collect::<Result<HashMap<String, String>>>()?;
+
+    // remove autovar.conf first, then scan all system entries and save
+    let tmpfiles_dir = tmprootfs_dfd
+        .open_dir(TMPFILESD)
+        .context("readdir {TMPFILESD}")?;
+
+    if tmpfiles_dir.try_exists(AUTOVAR_PATH)? {
+        tmpfiles_dir.remove_file(AUTOVAR_PATH)?;
+    }
+    let system_tmpfiles_entries = save_tmpfile_entries(&tmpfiles_dir)?
+        .map(|s| {
+            let entry = tmpfiles_entry_get_path(&s.as_str())?;
+            anyhow::Ok(entry.to_string())
+        })
+        .collect::<Result<HashSet<String>>>()?;
+
+    // remove duplicated entries in auto-generated tmpfiles.d,
+    // which are already in system tmpfiles
+    for key in system_tmpfiles_entries.into_iter() {
+        rpmostree_tmpfiles_entries.retain(|k, _value| k != &key);
+    }
+
+    {
+        // save the noduplicated entries
+        let mut entries = String::from("# This file was generated by rpm-ostree.\n");
+        for (_key, value) in rpmostree_tmpfiles_entries {
+            writeln!(entries, "{value}").unwrap();
+        }
+
+        let perms = Permissions::from_mode(0o644);
+        tmpfiles_dir.atomic_write_with_perms(&AUTOVAR_PATH, entries.as_bytes(), perms)?;
+    }
+    Ok(())
+}
+
+// #[context("Scan all tmpfiles conf and save entries")]
+fn save_tmpfile_entries(tmpfiles_dir: &Dir) -> Result<impl Iterator<Item = String>> {
+    let entries = tmpfiles_dir
+        .entries()?
+        .filter_map(|name| {
+            let name = name.unwrap().file_name();
+            if let Some(extension) = Path::new(&name).extension() {
+                if extension != "conf" {
+                    return None;
+                }
+            } else {
+                return None;
+            }
+            Some(
+                tmpfiles_dir
+                    .read_to_string(name)
+                    .unwrap()
```

Comment:
I'd like to eventually try to remove this `unwrap()` in favor of `?` but...trying this it gets really messy.  We can deal with it later.

Reply:
Sorry to forget about this, will change soon.

Reply:
Update `.ok()?` as the compiler suggests. Thanks for the comment.

Reply:
That changes from crashing on errors to ignoring them; neither are ideal.

I looked briefly at trying to fix this but it's nontrivial, and we can address it later.

Reply:
OK, thanks!